ecs-agent/acs/session/task_manifest_responder_test.go

```diff
@@ -0,0 +1,149 @@
+				// We expect only one task marked for termination.
+				require.Equal(t, len(stopVerification.StopCandidates), 1)
+				require.Equal(t, aws.StringValue(stopVerification.StopCandidates[0].TaskArn), testconst.TaskARN)
+			}
+		}
+		return nil
+	}
+
+	mockMetrics := mock_metrics.NewMockEntryFactory(ctrl)
+	mockComparer := mock_session.NewMockTaskComparer(ctrl)
+	mockIDA := mock_session.NewMockManifestMessageIDAccessor(ctrl)
+	mockSNA := mock_session.NewMockSequenceNumberAccessor(ctrl)
+
+	testTaskManifestResponder := NewTaskManifestResponder(mockComparer, mockSNA, mockIDA, mockMetrics, testResponseSender)
+
+	// Expect to get the latest sequence number for stale check, then set it as we expect.
+	mockSNA.EXPECT().GetLatestSequenceNumber()
+	mockSNA.EXPECT().SetLatestSequenceNumber(testManifest).Return(nil)
+
+	// After we get the latest sequence number, we will expect to set the latest message ID.
+	mockIDA.EXPECT().SetMessageID(testconst.MessageID).Return(nil)
+
+	// Test Metric publication when our responder starts and finishes message processing.
+	mockEntry := mock_metrics.NewMockEntry(ctrl)
+	mockMetrics.EXPECT().New(metrics.TaskManifestHandlingDuration).Return(mockEntry)
+	mockEntry.EXPECT().Done(nil).AnyTimes()
+
+	// When the client queries tasks to be stopped, inject this task to be stopped.
+	mockComparer.EXPECT().CompareRunningTasksOnInstanceWithManifest(testManifest).Return(
+		[]*ecsacs.TaskIdentifier{
+			{DesiredStatus: aws.String("STOPPED"), TaskArn: aws.String(testconst.TaskARN)},
+		},
+		nil,
+	)
+
+	// Handle the task manifest message update.
+	handleManifestMessage := testTaskManifestResponder.HandlerFunc().(func(*ecsacs.TaskManifestMessage))
+	handleManifestMessage(testManifest)
+}
+
+// TestTaskManifestStaleMessage will make sure the manifest can update, retrieve and catch stale
+// Manifest sequence numbers appropriately.
+func TestTaskManifestStaleMessage(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	testResponseSender := func(response interface{}) error {
+		return nil
+	}
+
+	mockComparer := mock_session.NewMockTaskComparer(ctrl)
+	mockIDA := mock_session.NewMockManifestMessageIDAccessor(ctrl)
+	mockSNA := mock_session.NewMockSequenceNumberAccessor(ctrl)
+
+	testTaskManifestResponder := NewTaskManifestResponder(mockComparer, mockSNA, mockIDA, metrics.NewNopEntryFactory(), testResponseSender)
+
+	// Create test task manifest.
+	testManifest := setupTestManifestMessage()
+
+	// Set up a new manifest with a stale number and distinct message ID.
+	newManifest := &ecsacs.TaskManifestMessage{
+		ClusterArn:           aws.String(testconst.ClusterName),
+		ContainerInstanceArn: aws.String(testconst.ContainerInstanceARN),
+		MessageId:            aws.String("456"),
+		Tasks:                []*ecsacs.TaskIdentifier{},
+		Timeline:             aws.Int64(EndingTaskManifestSequenceNumber),
+	}
+
+	// Expect GetLatestSequenceNumber once in handleManifestMessage, allowing it to pass
```

Comment:
> Expect GetLatestSequenceNumber once

Why are we using `.AnyTimes()` here then?

We do have two calls to `handleManifestMessage`, but I wouldn't expect `GetLatestSequenceNumber` to return the value held in `EndingTaskManifestSequenceNumber` const on the second call.

Reply:
Can update to .Times(2), but this coerces the return instead of declaring what it should be iiuc. Will also add in an expect for the startingtaskmanifestsequencenumber that we expect to compare against in the second time.

Reply:
> Can update to .Times(2), but this coerces the return instead of declaring what it should be iiuc

I'm not sure I understand - what is the difference between coercing the return and declaring what it should be? In any case, the value held in `EndingTaskManifestSequenceNumber` const shouldn't be returned on the second call (or any subsequent calls) in the context of this test, right?

Reply:
No, the first time we call GetLatestSequenceNumber I do not know what it is checking against(likely 0). We will know that it passes though by coercing it to return `StaleTaskManifestSequenceNumber` because 10<=15. The second time we call GetLatestSequenceNumber it will check against StartingTaskManifestSequenceNumber, so if you want to be explicit then we can declare those two function calls.

My understanding is that `.EXPECT().<function-call>.Return(xyz)` will return xyz when it is called, despite what it will do in a live environment, because mocks do not hold state. When I say coerce a return, I mean to tell <function-call> to return xyz. Declaring what it should be is the misconception that `EXPECT().<function-call>.Return(xyz)` returns xyz naturally in its lifecycle.

Reply:
This is resolved as of the most recent revision - the value that was held in`EndingTaskManifestSequenceNumber` const is no longer returned on the second call.